WebDriver Bidi: implement context created event and destroyed

AnujPanta1 opened the pull request:
#### c6598da6cf2a095929b14ae38e135f9a08ef9fcd
<pre>
WebDriver Bidi: implement context created event and destroyed
<a href="https://bugs.webkit.org/show_bug.cgi?id=297401">https://bugs.webkit.org/show_bug.cgi?id=297401</a>
<a href="https://rdar.apple.com/158325731">rdar://158325731</a>

Reviewed by BJ Burg.

Added contextCreated and contextDestroyed events to track browsing context life cycles.
Events are emitted at key points in page and frame creation/destruction to provide proper
information like parent relationship and opener details. Hooking into didCreatePage()
and didCreateSubFrame() for creation events. We recursively emit events for existing frame
hierarchies with getAllFrameTrees(). Destruction events are triggered from disconnect() for
frames and willClosePage() for pages. A key note is cleanup logic about removing context
handles from the m_handleWebPageMap was moved to the new destruction methods.

* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::WebAutomationSession::closeBrowsingContext):
(WebKit::WebAutomationSession::didCreatePage):
(WebKit::WebAutomationSession::emitContextCreatedEvent):
(WebKit::getClientWindowAndUserContext):
(WebKit::WebAutomationSession::getOpenerPage):
(WebKit::WebAutomationSession::didCreateFrame):
(WebKit::WebAutomationSession::willDestroyFrame):
(WebKit::WebAutomationSession::contextCreatedForFrame):
(WebKit::WebAutomationSession::recursivelyEmitContextCreatedEvent):
(WebKit::WebAutomationSession::contextDestroyedForPage):
(WebKit::WebAutomationSession::contextDestroyedForFrame):
(WebKit::WebAutomationSession::willClosePage):
* Source/WebKit/UIProcess/Automation/WebAutomationSession.h:
* Source/WebKit/UIProcess/Automation/protocol/BidiBrowsingContext.json:
* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::disconnect):
(WebKit::WebFrameProxy::isConnected const):
(WebKit::WebFrameProxy::didCreateSubframe):
* Source/WebKit/UIProcess/WebFrameProxy.h:

Canonical link: <a href="https://commits.webkit.org/302865@main">https://commits.webkit.org/302865@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7d967b9d5a4b1693317acdb0a23126c7cb1c2cd8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/129450 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/1707 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/40289 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/136827 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/80869 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/d129fa07-105a-4552-9677-a4d303dbb1ad) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/1639 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/1583 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/98591 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/66481 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/264b530f-60e1-46a8-8070-e143595e162f) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/132397 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/1281 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/115946 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/79243 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/5db0c576-6e03-4f59-9241-4581548900fa) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/1201 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/80104 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/109706 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/34574 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/139303 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/1497 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/1438 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/107117 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/1539 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/112287 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/106961 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27444 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/1220 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/30807 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/54169 "The change is no longer eligible for processing.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/1568 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/64931 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/1386 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/1422 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/1490 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

burg commented on Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
```diff
@@ -955,6 +952,7 @@ void WebAutomationSession::wheelEventsFlushedForPage(const WebPageProxy& page)
 void WebAutomationSession::didCreatePage(WebPageProxy& page)
 {
     m_bidiProcessor->browserAgent().didCreatePage(page);
+    contextCreatedForPage(page);
```
Naming: `emitContextCreatedEvent`

burg commented on Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
```diff
@@ -990,13 +988,138 @@ void WebAutomationSession::fragmentNavigatedForFrame(const WebFrameProxy& frame,
+                return;
+
+            for (auto& tree : trees)
+                recursivelyEmitContextCreated(tree, std::nullopt);
```
`recursivelyEmitContextCreatedEvent`

burg commented on Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
```diff
@@ -990,13 +988,138 @@ void WebAutomationSession::fragmentNavigatedForFrame(const WebFrameProxy& frame,
+    }
+}
+
+std::pair<String, String> WebAutomationSession::getClientWindowAndUserContext(const WebPageProxy& page)
```
This can be a static helper function rather than an instance method, if it's not being used outside the class.

burg commented on Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
```diff
@@ -990,13 +988,138 @@ void WebAutomationSession::fragmentNavigatedForFrame(const WebFrameProxy& frame,
+std::pair<String, String> WebAutomationSession::getClientWindowAndUserContext(const WebPageProxy& page)
+{
+    String clientWindow = makeString(page.identifier().toUInt64());
+    String userContext = "default"_s;
```
Nit: add FIXME for reporting user context

burg commented on Source/WebKit/UIProcess/WebFrameProxy.cpp:
```diff
@@ -53,6 +53,9 @@
 #include "WebProcessPool.h"
 #include "WebsiteDataStore.h"
 #include "WebsitePoliciesData.h"
+#if ENABLE(WEBDRIVER_BIDI)
```
It's OK, but you need to move the conditional include to be below all the unconditional includes.

burg commented on Source/WebKit/UIProcess/Automation/WebAutomationSession.h:
```diff
@@ -171,6 +172,13 @@ friend class WebDriverBidiProcessor;
+    void contextCreatedForFrame(const WebFrameProxy&);
+    void contextDestroyedForPage(const WebPageProxy&);
+    void contextDestroyedForFrame(const WebFrameProxy&);
+    void recursivelyEmitContextCreated(const FrameTreeNodeData&, std::optional<String>&& parentContext);
+    std::pair<String, String> getClientWindowAndUserContext(const WebPageProxy&);
+    RefPtr<WebPageProxy> getOpenerPage(const WebPageProxy&);
```
These last three don't need to be in the header file, they can be private to the .cpp file.

burg commented on Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
```diff
@@ -990,13 +988,138 @@ void WebAutomationSession::fragmentNavigatedForFrame(const WebFrameProxy& frame,
+        userContext = contextId;
+    }
+
+    m_bidiProcessor->browsingContextDomainNotifier().contextCreated(contextHandle, url, "null"_s, parentHandle, JSON::ArrayOf<Inspector::Protocol::BidiBrowsingContext::Info>::create(), clientWindow, userContext);
```
I'm not sure that we should be passing `null` here, which argument is that? Okay, I looked it up.. please use `let originalOpener` so it has a name.

I'm a little concerned that setting `"null"` won't actually set the field to `null` in the JSON payload, it will instead set a string. Have you looked into what it does? We might need to come up with a new solution for setting null fields in events, if it doesn't work already.

lauromoura replied:
FTR, this support can be added in a separate PR, keeping this one focused on the `contextCreated/Destroyed` events.

burg commented on Source/WebKit/UIProcess/WebFrameProxy.cpp:
```diff
@@ -449,8 +452,15 @@ void WebFrameProxy::collapseSelection()
+#if ENABLE(WEBDRIVER_BIDI)
+        if (RefPtr page = m_page.get()) {
+            if (RefPtr session = page->activeAutomationSession())
+                session->contextDestroyedForFrame(*this);
```
How is this related to the existing `WebAutomationSession::didDestroyFrame` which does the `m_frameHandleMap` maintenance?

burg commented on Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
```diff
@@ -990,13 +988,138 @@ void WebAutomationSession::fragmentNavigatedForFrame(const WebFrameProxy& frame,
+
+    m_bidiProcessor->browsingContextDomainNotifier().contextDestroyed(contextHandle, url, originalOpener, parentContext, JSON::ArrayOf<Inspector::Protocol::BidiBrowsingContext::Info>::create(), clientWindow, userContext);
+
+    m_handleWebPageMap.remove(contextHandle);
```
I think it's OK to do this map maintenance here, but then we need to ensure that these two new instrumentation methods are called unconditionally. As-is, in this patch we remove some bookkeeping but don't add it back if `!ENABLE(WEBDRIVER_BIDI)`

burg reviewed (changes requested):
Marking request changes while we figure out if changes to the bookkeeping scheme are correct.

burg commented:
Please wrap changelog to 100 characters or so. It's really hard to read right now on one line.

Please address SaferCPP failures.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/fa3222bc4bf557b4872e7a2f49d240981e519c9c. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50581

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/604c6107882b727942b11213d226bbf7da4f882f. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50581

burg commented on Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
```diff
@@ -990,13 +988,160 @@ void WebAutomationSession::fragmentNavigatedForFrame(const WebFrameProxy& frame,
+    if (RefPtr parentFrame = frame.parentFrame())
+        parentHandle = handleForWebFrameProxy(*parentFrame);
+
+    String clientWindow = "unknown-window"_s;
```
Add the FIXME comment here as well

burg commented on Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
```diff
@@ -981,6 +978,7 @@ void WebAutomationSession::wheelEventsFlushedForPage(const WebPageProxy& page)
 void WebAutomationSession::didCreatePage(WebPageProxy& page)
 {
     m_bidiProcessor->browserAgent().didCreatePage(page);
+    emitContextCreatedEvent(page);
```
I'm not understanding how this recursively emitting context created is supposed to work. This is the only place it gets called, and at this point we'd never have any subframes yet. should `contextCreatedForFrame` be calling into the recursively emit events too?

webkit-early-warning-system commented:
EWS run on current version of this PR (hash https://github.com/WebKit/WebKit/commit/7d967b9d5a4b1693317acdb0a23126c7cb1c2cd8)<details>

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/129450 "Passed style check") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/1707 "Build is in progress. Recent messages:OS: Tahoe (26.0), Xcode: 26.0.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/40289 "Build is in progress. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4; Running apply-patch; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/136827 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [⏳ 🛠 win ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") | ⏳ 🛠 ios-apple 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/1639 "Build is in progress. Recent messages:OS: Tahoe (26.0), Xcode: 26.0.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/1583 "Build is in progress. Recent messages:OS: Tahoe (26.0), Xcode: 26.0.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/5/builds/136827 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [⏳ 🧪 win-tests ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/132397 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/155/builds/1639 "Build is in progress. Recent messages:OS: Tahoe (26.0), Xcode: 26.0.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-mac](https://ews-build.webkit.org/#/builders/138/builds/40289 "Build is in progress. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4; Running apply-patch; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-wpe](https://ews-build.webkit.org/#/builders/5/builds/136827 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | | ⏳ 🛠 vision-apple 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/155/builds/1639 "Build is in progress. Recent messages:OS: Tahoe (26.0), Xcode: 26.0.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/138/builds/40289 "Build is in progress. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4; Running apply-patch; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/80104 "Build is in progress. Recent messages:Printed configuration") | | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-ios](https://ews-build.webkit.org/#/builders/155/builds/1639 "Build is in progress. Recent messages:OS: Tahoe (26.0), Xcode: 26.0.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/138/builds/40289 "Build is in progress. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4; Running apply-patch; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/139303 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/1497 "Build is in progress. Recent messages:OS: Tahoe (26.0), Xcode: 26.0.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/156/builds/1583 "Build is in progress. Recent messages:OS: Tahoe (26.0), Xcode: 26.0.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/2/builds/139303 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/1539 "Build is in progress. Recent messages:OS: Tahoe (26.0), Xcode: 26.0.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/138/builds/40289 "Build is in progress. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4; Running apply-patch; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-gtk](https://ews-build.webkit.org/#/builders/2/builds/139303 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-26-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/138/builds/40289 "Build is in progress. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4; Running apply-patch; Checked out pull request; Running compile-webkit") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/54169 "The change is no longer eligible for processing.") | | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/1568 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/64931 "Build is in progress. Recent messages:OS: Sequoia (15.5), Xcode: 16.4; Checked out pull request; Running scan-build") | | | 
| | [⏳ 🛠 tv-sim ](https://ews-build.webkit.org/#/builders/tvOS-26-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | | | | 
| | [⏳ 🛠 watch ](https://ews-build.webkit.org/#/builders/watchOS-26-Build-EWS "Waiting in queue, processing has not started yet") | | | | 
| | [⏳ 🛠 watch-sim ](https://ews-build.webkit.org/#/builders/watchOS-26-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | | | | </details>

AnujPanta1 committed:
WebDriver Bidi: implement context created event and destroyed
https://bugs.webkit.org/show_bug.cgi?id=297401
rdar://158325731

Reviewed by BJ Burg.

Added contextCreated and contextDestroyed events to track browsing context life cycles.
Events are emitted at key points in page and frame creation/destruction to provide proper
information like parent relationship and opener details. Hooking into didCreatePage()
and didCreateSubFrame() for creation events. We recursively emit events for existing frame
hierarchies with getAllFrameTrees(). Destruction events are triggered from disconnect() for
frames and willClosePage() for pages. A key note is cleanup logic about removing context
handles from the m_handleWebPageMap was moved to the new destruction methods.

* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::WebAutomationSession::closeBrowsingContext):
(WebKit::WebAutomationSession::didCreatePage):
(WebKit::WebAutomationSession::emitContextCreatedEvent):
(WebKit::getClientWindowAndUserContext):
(WebKit::WebAutomationSession::getOpenerPage):
(WebKit::WebAutomationSession::didCreateFrame):
(WebKit::WebAutomationSession::willDestroyFrame):
(WebKit::WebAutomationSession::contextCreatedForFrame):
(WebKit::WebAutomationSession::recursivelyEmitContextCreatedEvent):
(WebKit::WebAutomationSession::contextDestroyedForPage):
(WebKit::WebAutomationSession::contextDestroyedForFrame):
(WebKit::WebAutomationSession::willClosePage):
* Source/WebKit/UIProcess/Automation/WebAutomationSession.h:
* Source/WebKit/UIProcess/Automation/protocol/BidiBrowsingContext.json:
* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::disconnect):
(WebKit::WebFrameProxy::isConnected const):
(WebKit::WebFrameProxy::didCreateSubframe):
* Source/WebKit/UIProcess/WebFrameProxy.h:

Canonical link: https://commits.webkit.org/302865@main

webkit-commit-queue commented:
Committed 302865@main (c6598da6cf2a): <https://commits.webkit.org/302865@main>

Reviewed commits have been landed. Closing PR #50581 and removing active labels.
